base/src/commonMain/kotlin/flow/SequentialMutableSharedFlow.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2022 Splendo Consulting B.V. The Netherlands
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ */
+
+package com.splendo.kaluga.base.flow
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * A [MutableSharedFlow] that can be forced to emit in a thread-safe sequential way
+ */
+interface SequentialMutableSharedFlow<T> : MutableSharedFlow<T> {
+
+    /**
+     * Tries to call [tryEmit] but if this fails will launch and [emit]
+     * The call order of this method is preserved in a thread-safe fashion
+     * @return `true` if [tryEmit] succeeded, false otherwise
+     */
+    fun tryEmitOrLaunchAndEmit(value: T): Boolean
+}
+
+/**
+ * Creates a new [SequentialMutableSharedFlow]
+ * @param replay the number of values replayed to new subscriber (cannot be negative, defaults to zero).
+ * @param extraBufferCapacity the number of values buffered in addition to replay.
+ * [SequentialMutableSharedFlow.emit] or [SequentialMutableSharedFlow.tryEmitOrLaunchAndEmit] does not suspend while there is a buffer space remaining
+ * (optional, cannot be negative. defaults to zero).
+ * @param coroutineScope The [coroutineScope] responsible for emitting when [SequentialMutableSharedFlow.tryEmitOrLaunchAndEmit] returns `false`.
+ */
+fun <T> SequentialMutableSharedFlow(replay: Int = 0, extraBufferCapacity: Int = 0, coroutineScope: CoroutineScope): SequentialMutableSharedFlow<T> = SequentialMutableSharedFlowImpl(replay, extraBufferCapacity, coroutineScope)
+
+private class SequentialMutableSharedFlowImpl<T>(replay: Int = 0, extraBufferCapacity: Int = 0, coroutineScope: CoroutineScope) : SequentialMutableSharedFlow<T>, CoroutineScope by coroutineScope {
```

Comment:
I'd reformat this long lines. It will helps to review and edit as well

Reply:
class is now gone